backtracking/letter_combinations_of_a_phone_number.cpp

```diff
@@ -0,0 +1,150 @@
+/**
+ * @file
+ * @brief Generates letter combinations for given digits using backtracking.
+ *
+ * @details
+ * This code generates all possible letter combinations corresponding to a given
+ * string of digits. It uses the concept of backtracking to explore all possible
+ * combinations of letters for each digit.
+ *
+ * The function letterCombinations takes a string of digits as input and returns
+ * a vector containing all the generated letter combinations. Each digit is
+ * mapped to a set of letters, and the function backtracks to explore all
+ * possible combinations by appending one letter at a time for each digit.
+ *
+ * The main test function demonstrates the correctness of letterCombinations by
+ * running multiple test cases and checking the output against the expected
+ * results. The expected vectors are now initialized one by one for better
+ * clarity.
+ *
+ * @author [Saurabh Kushwaha](https://github.com/sOR-o)
+ */
+
+#include <array>     /// for std::array
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+#include <vector>    /// for std::vector
+
+/**
+ * @namespace backtracking
+ * @brief Backtracking algorithms
+ */
+namespace backtracking {
+/**
+ * @namespace letter_combinations
+ * @brief Functions for generating letter combinations from digits using
+ * backtracking.
+ */
+namespace letter_combinations {
+/**
+ * @brief Recursive utility function to generate letter combinations.
+ * @param digit The input digit string to generate combinations for.
+ * @param output The current output combination being formed.
+ * @param index The index of the current digit being processed.
+ * @param ans Vector to store the generated combinations.
+ * @param mpp Array containing the mapping of digits to letters.
+ */
+void solve(const std::string& digit, std::string& output, int index,
+           std::vector<std::string>& ans,
+           const std::array<std::string, 10>& mpp) {
+    if (index >= digit.length()) {
+        ans.push_back(output);
+        return;
+    }
+
+    int num = digit[index] - '0';
+    std::string value = mpp[num];
+
+    for (int i = 0; i < value.size(); i++) {
+        output.push_back(value[i]);
+        solve(digit, output, index + 1, ans, mpp);
+        output.pop_back();
+    }
+}
+
+/**
+ * @brief Generates letter combinations for a given digit string.
+ * @param digit The input digit string to generate combinations for.
+ * @returns A vector containing all the generated letter combinations.
+ */
+std::vector<std::string> letterCombinations(const std::string& digit) {
+    std::vector<std::string> ans;
+
+    if (digit.empty()) {
+        return ans;
+    }
+
+    std::string output = "";
+    int index = 0;
+
+    std::array<std::string, 10> mpp = {"",    "",    "abc",  "def", "ghi",
+                                       "jkl", "mno", "pqrs", "tuv", "wxyz"};
+    solve(digit, output, index, ans, mpp);
+    return ans;
+}
```

Comment:
```suggestion
std::vector<std::string> letterCombinations(const std::string& digit) {
    std::vector<std::string> ans;

    if (digit.empty()) {
        return ans;
    }

    const std::array<std::string, 10> mpp = {
        "", "", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
    std::string output = "";
    solve(digit, output, 0, ans, mpp);
    return ans;
}
```
Formatting was done by clang-format. I suggest to:
- mark `mpp` as `const`,
- remove the `index` variable.